vendors/circom/circomlib/r1cs/r1cs.h

```diff
@@ -0,0 +1,363 @@
+#ifndef VENDORS_CIRCOM_CIRCOMLIB_R1CS_R1CS_H_
+#define VENDORS_CIRCOM_CIRCOMLIB_R1CS_R1CS_H_
+
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "tachyon/base/buffer/copyable.h"
+#include "tachyon/base/buffer/endian_auto_reset.h"
+#include "tachyon/base/logging.h"
+#include "tachyon/base/strings/string_util.h"
+#include "tachyon/math/base/big_int.h"
+
+namespace tachyon {
+namespace circom {
+namespace v1 {
+
+struct R1CS;
+
+}  // namespace v1
+
+struct R1CS {
+  virtual ~R1CS() = default;
+
+  virtual uint32_t GetVersion() const = 0;
+
+  virtual v1::R1CS* ToV1() { return nullptr; }
+};
+
+constexpr char kR1CSMagic[4] = {'r', '1', 'c', 's'};
+
+struct PrimeField {
+  std::vector<uint8_t> bytes;
+
+  bool operator==(const PrimeField& other) const {
+    return bytes == other.bytes;
+  }
+  bool operator!=(const PrimeField& other) const {
+    return bytes != other.bytes;
+  }
+
+  template <size_t N>
+  math::BigInt<N> ToBigInt() const {
+    return math::BigInt<N>::FromBytesLE(bytes);
+  }
+
+  template <size_t N>
+  static PrimeField FromBigInt(const math::BigInt<N>& big_int) {
+    std::array<uint8_t, N * 8> bytes = big_int.ToBytesLE();
+    return {{bytes.begin(), bytes.end()}};
+  }
+
+  std::string ToString() const;
+};
+
+namespace v1 {
+
+enum class SectionType : uint32_t {
+  kHeader = 0x1,
+  kConstraints = 0x2,
+  kWire2LabelIdMap = 0x3,
+  kCustomGatesList = 0x4,
+  kCustomGatesApplication = 0x5,
+};
+
+std::string_view SectionTypeToString(SectionType type);
+
+struct HeaderSection {
+  PrimeField modulus;
+  // Total number of wires including ONE signal (Index 0).
+  uint32_t num_wires;
+  // Total number of wires public output wires. They should be starting at
+  // idx 1.
+  uint32_t num_public_outputs;
+  // Total number of wires public input wires. They should be starting just
+  // after the public output.
+  uint32_t num_public_inputs;
+  // Total number of wires private input wires. They should be starting just
```

Comment:
Please remove duplicate words.
```suggestion
  // Total number of private input wires. They should be starting just
```